zkevm-circuits/src/mpt_circuit/branch/branch_key.rs

```diff
@@ -108,358 +128,124 @@ pub(crate) struct BranchKeyConfig<F> {
+            let modified_index = a!(ctx.branch.modified_index, rot_first_child);
+            let key_rlc =
+                DataTransition::new_with_rot(meta, key.rlc, rot_first_child_prev, rot_first_child);
+            let key_mult =
+                DataTransition::new_with_rot(meta, key.mult, rot_first_child_prev, rot_first_child);
+
+            ifx! {f!(position_cols.q_not_first_ext_c), a!(ctx.branch.is_extension_node_c) => {
+                let selectors = [branch.is_key_odd(), branch.is_key_even()];
+                // Selectors need to be boolean.
+                for selector in selectors.iter() {
+                    require!(selector => bool);
+                }
+                // One selector needs to be enabled.
+                require!(sum::expr(&selectors) => 1);
+
+                // Get the previous values from the previous branch. In the first level use initial values.
+                let after_first_level = not_first_level.expr() * not!(branch.is_below_account(meta));
+                let (key_rlc_prev, key_mult_prev) = ifx!{after_first_level => {
+                    (key_rlc.prev(), key_mult.prev())
+                } elsex {
+                    (0.expr(), 1.expr())
+                }};
+
+                // Calculate the extension node key RLC when in an extension node
+                let key_rlc_post_ext = ifx!{branch.is_extension() => {
+                    let key_rlc_ext = DataTransition::new(meta, key.rlc);
+                    // Extension key rlc
+                    let ext_key_rlc = key_rlc_prev.expr() + branch.ext_key_rlc(meta, &mut cb.base, key_mult_prev.expr());
+                    // Currently, the extension node S and extension node C both have the same key RLC -
+                    // however, sometimes extension node can be replaced by a shorter extension node
+                    // (in terms of nibbles), this is still to be implemented.
+                    // TODO: extension nodes of different nibbles length
+                    require!(key_rlc_ext => key_rlc_ext.prev());
+                    // Store it
+                    require!(key_rlc_ext => ext_key_rlc);
+                    ext_key_rlc.expr()
+                } elsex {
+                    key_rlc_prev.expr()
+                }};
+
+                // Get the length of the key
+                // If the resulting key is odd then subtract 1 so we still use the previous multiplier.
+                // TODO(Brecht): is_long_even_c1 needs -1 while it should be even?
+                let key_num_bytes_for_mult = ifx!{branch.is_extension() => {
+                    let key_len = branch.ext_key_len(meta, -1);
+                    matchx! {
+                        branch.is_key_post_ext_even() - branch.is_long_even_c1.expr() => key_len.expr(),
+                        branch.is_key_post_ext_odd() + branch.is_long_even_c1.expr() => key_len.expr() - 1.expr(),
+                    }
```

Comment:
Seems like `is_key_post_ext_even` actually returns `is_previous_branch_c16` and `is_key_post_ext_odd` returns `is_previous_branch_c1`.

Whether the number of nibbles is even or odd after the extension node is actually given by `is_c16` and `is_c1`.

So, for example `is_short_c16` means that there are even nibbles up to this branch (because of `c16`) and there is one nibble in the extension node above the branch (because of `short`), so the previous branch is `c16` (because the nibble in the extension node is `c1`).

It seems like `branch.is_key_post_ext_even() - branch.is_long_even_c1.expr()` can be rephrased into something like `branch.is_previous_branch_c16() - branch.is_long_even_c1.expr()`.

Similarly, `branch.is_key_post_ext_odd() - branch.is_long_even_c1.expr()` can be rephrased into `branch.is_previous_branch_c1() - branch.is_long_even_c1.expr()`.

It’s kind of difficult to visualize the two conditions, but I think from the examples below it can be seen why sometimes it is `key_len` and sometimes `key_len - 1`. For example, if in the first two cases the number of nibbles is 2 and 3 respectively, then the mult_diff needs to be `r` in the first case and `r^2` in the second case.

- previous branch c16 & long even & c1: 
  - first condition: 1 - 1 = 0
  - second condition: 0 + 1 = 1

- previous branch c16 & long odd & c16:
  - first condition: 1 - 0 = 1
  - second condition: 0 + 0 = 0

- previous branch c16 & long even & c16 (can’t happen)

- previous branch c1 & long even & c16:
  - first condition: 0 - 0 = 0
  - second condition: 1 + 0 = 1

- previous branch c1 & long even & c1 (can’t happen)

- previous branch c1 & long odd & c16 (can’t happen)
  
- previous branch c1 & long odd & c1:
  - first condition: 0 - 0 = 0
  - second condition: 1 + 0 = 1